File: BOJ_Python/스택, 큐/BOJ_17298_오큰수.py
import sys

n = int(sys.stdin.readline())
num = list(map(int, sys.stdin.readline().split()))
ans = [-1] * n
stack = [0]

for i in range(1, n):
    while stack and num[stack[-1]] < num[i]:
        ans[stack.pop()] = num[i]
    stack.append(i)

print(*ans)
# print(ans)

File: BOJ_Swift/스택, 큐, 덱/BOJ_17298_오큰수.swift
//
//  main.swift
//  Algorithm
//
//  Created by GYURI PARK on 2023/11/28.
//

import Foundation

let n = Int(readLine()!)!
let num = Array(readLine()!.split(separator: " ").map{Int($0)!})
var ans = [Int]()
// var ans = Array(repeating: -1, count: n)
var stack = [Int]()
stack.append(0)

for _ in 0..<n {
    ans.append(-1)
}

for i in 1..<n {
    while !stack.isEmpty && num[stack.last!] < num[i] {
        ans[stack.popLast()!] = num[i]
    }
    stack.append(i)
}

print(ans.map{String($0)}.joined(separator: " "))
